File: apps/ruby/test/chapter02_test.rb
# frozen_string_literal: true

require "test_helper"

class Chapter02Test < Minitest::Test
  C = GettingStartedMl::Chapter02

  def row(cells)
    C::Row.new(cells)
  end

  def test_数値の列を読む
    assert_in_delta 5.1, row("がく片長さ" => "5.1").number("がく片長さ")
  end

  def test_空欄の列は欠損値になる
    cells = row("がく片長さ" => "")

    assert_nil cells.number("がく片長さ")
    assert cells.missing?("がく片長さ")
  end

  def test_数値として読めない列は失敗する
    error = assert_raises(ArgumentError) { row("がく片長さ" => "たくさん").number("がく片長さ") }

    assert_equal "がく片長さ を数値として読めません: たくさん", error.message
  end

  def test_列が無ければ失敗する
    error = assert_raises(KeyError) { row("がく片長さ" => "5.1").text("花弁幅") }

    assert_equal "列がありません: 花弁幅", error.message
  end

  def test_文字列の列を読む
    assert_equal "Iris-setosa", row("種類" => "Iris-setosa").text("種類")
  end

  def test_列ごとに欠損値を数える
    table = C::Table.new(
      columns: %w[がく片長さ 種類],
      rows: [
        row("がく片長さ" => "5.1", "種類" => "Iris-setosa"),
        row("がく片長さ" => "", "種類" => "Iris-setosa"),
        row("がく片長さ" => "", "種類" => "Iris-virginica")
      ]
    )

    assert_equal({ "がく片長さ" => 2, "種類" => 0 }, table.count_missing)
  end

  def test_欠損値を除いて列ごとの平均値を求める
    rows = [row("がく片長さ" => "1.0"), row("がく片長さ" => ""), row("がく片長さ" => "3.0")]

    assert_equal({ "がく片長さ" => 2.0 }, C.column_means(rows, ["がく片長さ"]))
  end

  def test_値がすべて空欄なら平均値を求められない
    error = assert_raises(ArgumentError) { C.column_means([row("がく片長さ" => "")], ["がく片長さ"]) }

    assert_equal "値がすべて空欄です: がく片長さ", error.message
  end

  def test_欠損値を平均値で補完する
    rows = [row("がく片長さ" => "1.0"), row("がく片長さ" => "")]

    filled = C.fill_missing(rows, ["がく片長さ"], { "がく片長さ" => 2.0 })

    assert_equal [C::Features.new(columns: ["がく片長さ"], values: [1.0]),
                  C::Features.new(columns: ["がく片長さ"], values: [2.0])], filled
  end

  def test_補完する値が無ければ失敗する
    error = assert_raises(KeyError) { C.fill_missing([row("がく片長さ" => "")], ["がく片長さ"], {}) }

    assert_equal "補完する値がありません: がく片長さ", error.message
  end

  def test_列名で特徴量の値を読む
    features = C::Features.new(columns: %w[がく片長さ 花弁幅], values: [5.1, 0.2])

    assert_in_delta 0.2, features.value("花弁幅")
    error = assert_raises(KeyError) { features.value("種類") }
    assert_equal "列がありません: 種類", error.message
  end

  def test_列名と値の数が合わなければ特徴量を作れない
    error = assert_raises(ArgumentError) { C::Features.new(columns: ["がく片長さ"], values: [5.1, 0.2]) }

    assert_equal "件数が違います: 1 と 2", error.message
  end

  def test_同じシードなら同じ並びになる
    items = (0..9).to_a

    assert_equal C.shuffle(items, 0), C.shuffle(items, 0)
    refute_equal C.shuffle(items, 0), C.shuffle(items, 1)
  end

  def test_並べ替えの並びはほかの言語版と違う
    items = (0..9).to_a

    # Rust 版は [9 3 6 4 8 1 5 2 0 7]、Java 版は [4 8 9 6 3 5 2 1 7 0]、Go 版は [6 8 2 3 7 5 9 1 0 4]
    assert_equal [2, 8, 4, 9, 1, 6, 7, 3, 0, 5], C.shuffle(items, 0)
  end

  def test_並べ替えても要素は変わらず元の配列も変わらない
    items = (0..9).to_a

    assert_equal items, C.shuffle(items, 0).sort
    assert_equal (0..9).to_a, items
  end

  def test_テストデータの割合で分ける
    x = (0..9).to_a

    split = C.split_train_test(x, x.map(&:to_s), test_size: 0.3, seed: 0)

    assert_equal [7, 3, 7, 3], [split.x_train.size, split.x_test.size, split.t_train.size, split.t_test.size]
  end

  def test_分けても特徴量と正解ラベルの対応は崩れない
    x = (0..9).to_a

    split = C.split_train_test(x, x.map(&:to_s), test_size: 0.3, seed: 0)

    assert_equal split.x_train.map(&:to_s), split.t_train
    assert_equal split.x_test.map(&:to_s), split.t_test
  end

  def test_特徴量と正解ラベルの件数が違えば分けられない
    error = assert_raises(ArgumentError) { C.split_train_test([1, 2], [1], test_size: 0.3, seed: 0) }

    assert_equal "件数が違います: 2 と 1", error.message
  end

  def test_正解ラベルの列を取り出して残りを特徴量の列にする
    table = C::Table.new(
      columns: %w[がく片長さ 種類],
      rows: [row("がく片長さ" => "5.1", "種類" => "Iris-setosa"), row("がく片長さ" => "6.0", "種類" => "Iris-virginica")]
    )

    columns, rows, labels = C.split_features_and_target(table, C::TARGET)

    assert_equal ["がく片長さ"], columns
    assert_equal 2, rows.size
    assert_equal %w[Iris-setosa Iris-virginica], labels
  end
end

File: apps/php/tests/Chapter02Test.php
<?php

declare(strict_types=1);

namespace GettingStartedMl\Tests;

use GettingStartedMl\Chapter02;
use GettingStartedMl\Csv;
use GettingStartedMl\Dataset;
use InvalidArgumentException;
use PHPUnit\Framework\Attributes\Group;
use PHPUnit\Framework\Attributes\TestDox;
use PHPUnit\Framework\TestCase;

final class Chapter02Test extends TestCase
{
    /** @var list<string> 後始末するファイル */
    private array $paths = [];

    private function csv(string $contents): string
    {
        $path = tempnam(sys_get_temp_dir(), 'iris') . '.csv';
        file_put_contents($path, $contents);
        $this->paths[] = $path;

        return $path;
    }

    protected function tearDown(): void
    {
        foreach ($this->paths as $path) {
            @unlink($path);
        }

        $this->paths = [];
    }

    #[TestDox('空欄は欠損値として数える')]
    public function test空欄は欠損値として数える(): void
    {
        $table = Csv::parseTable("a,b\n1,\n,2\n3,4\n");

        $this->assertSame(['a' => 1, 'b' => 1], Chapter02::countMissing($table));
    }

    #[TestDox('空欄の数値は null になる')]
    public function test空欄の数値はnullになる(): void
    {
        $this->assertNull(Chapter02::number(['a' => ' '], 'a'));
        $this->assertSame(1.5, Chapter02::number(['a' => '1.5'], 'a'));
    }

    #[TestDox('数値として読めない値があれば失敗する')]
    public function test数値として読めない値があれば失敗する(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('a を数値として読めません: 高い');

        Chapter02::number(['a' => '高い'], 'a');
    }

    #[TestDox('列が無ければ読めない')]
    public function test列が無ければ読めない(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('列がありません: b');

        Chapter02::text(['a' => '1'], 'b');
    }

    #[TestDox('欠損値を除いて平均値を求める')]
    public function test欠損値を除いて平均値を求める(): void
    {
        $rows = [['a' => '1'], ['a' => ''], ['a' => '3']];

        $this->assertEqualsWithDelta(['a' => 2.0], Chapter02::columnMeans($rows, ['a']), 1e-9);
    }

    #[TestDox('値がすべて空欄なら平均値を求められない')]
    public function test値がすべて空欄なら平均値を求められない(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('値がすべて空欄です: a');

        Chapter02::columnMeans([['a' => ''], ['a' => '']], ['a']);
    }

    #[TestDox('欠損値を指定した値で補完する')]
    public function test欠損値を指定した値で補完する(): void
    {
        $filled = Chapter02::fillMissing([['a' => '1'], ['a' => '']], ['a'], ['a' => 9.0]);

        $this->assertEqualsWithDelta([['a' => 1.0], ['a' => 9.0]], $filled, 1e-9);
    }

    #[TestDox('補完する値が無ければ失敗する')]
    public function test補完する値が無ければ失敗する(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('補完する値がありません: a');

        Chapter02::fillMissing([['a' => '']], ['a'], []);
    }

    #[TestDox('正解ラベルの列を特徴量から外す')]
    public function test正解ラベルの列を特徴量から外す(): void
    {
        $table = Csv::parseTable("a,種類\n1,setosa\n");

        [$columns, $rows, $labels] = Chapter02::splitFeaturesAndTarget($table, Chapter02::TARGET);

        $this->assertSame(['a'], $columns);
        $this->assertCount(1, $rows);
        $this->assertSame(['setosa'], $labels);
    }

    #[TestDox('テストデータの割合は切り上げる')]
    public function testテストデータの割合は切り上げる(): void
    {
        $x = array_map(static fn (int $i): array => ['a' => (string) $i], range(1, 10));
        $t = array_map(static fn (int $i): string => "t{$i}", range(1, 10));

        $split = Chapter02::splitTrainTest($x, $t, 0.25, 0);

        // 10 件の 25% は 2.5 なので、切り上げて 3 件がテストデータになる。
        $this->assertCount(7, $split['xTrain']);
        $this->assertCount(3, $split['xTest']);
        $this->assertCount(7, $split['tTrain']);
        $this->assertCount(3, $split['tTest']);
    }

    #[TestDox('件数が違えば分割できない')]
    public function test件数が違えば分割できない(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('件数が違います: 2 と 1');

        Chapter02::splitTrainTest([['a' => '1'], ['a' => '2']], ['t'], 0.5, 0);
    }

    #[TestDox('分割しても特徴量とラベルの組が崩れない')]
    public function test分割しても組が崩れない(): void
    {
        $x = array_map(static fn (int $i): array => ['a' => (string) $i], range(1, 10));
        $t = array_map(static fn (int $i): string => "t{$i}", range(1, 10));

        $split = Chapter02::splitTrainTest($x, $t, 0.3, 0);

        foreach ($split['xTrain'] as $i => $row) {
            $this->assertSame('t' . $row['a'], $split['tTrain'][$i]);
        }
    }

    #[TestDox('前処理のあとに空欄が残らない')]
    public function test前処理のあとに空欄が残らない(): void
    {
        $path = $this->csv("がく片長さ,種類\n1,setosa\n3,setosa\n,versicolor\n5,versicolor\n");

        $prepared = Chapter02::prepareIris($path, 0.25, 0);
        $values = array_merge(
            array_column($prepared['xTrain'], 'がく片長さ'),
            array_column($prepared['xTest'], 'がく片長さ'),
        );

        // 4 件すべてが数値になっている（補完されなかった行は元の値のまま）。
        $this->assertCount(4, $values);
        $this->assertSame(4, count(array_filter($values, is_finite(...))));
    }

    #[TestDox('補完に使う平均値は訓練データだけから求める')]
    public function test補完に使う平均値は訓練データだけから求める(): void
    {
        // 訓練データだけの平均（1 と 3 の平均＝2.0）で補完し、テストデータの 5 は混ぜない。
        $train = [['a' => '1'], ['a' => '3']];
        $means = Chapter02::columnMeans($train, ['a']);

        $this->assertEqualsWithDelta(['a' => 2.0], $means, 1e-9);
        $this->assertEqualsWithDelta(
            [['a' => 2.0]],
            Chapter02::fillMissing([['a' => '']], ['a'], $means),
            1e-9,
        );
    }

    #[Group('data')]
    #[TestDox('実データの前処理がほかの言語版と一致する')]
    public function test実データの前処理が一致する(): void
    {
        if (!Dataset::exists('iris.csv')) {
            $this->markTestSkipped('学習データがありません: ' . Dataset::path('iris.csv'));
        }

        $path = Dataset::path('iris.csv');
        $table = Chapter02::loadTable($path);

        $this->assertCount(150, $table->rows);
        $this->assertSame(
            ['がく片長さ' => 2, 'がく片幅' => 1, '花弁長さ' => 2, '花弁幅' => 2, '種類' => 0],
            Chapter02::countMissing($table),
        );

        $split = Chapter02::prepareIris($path, 0.3, 0);

        $this->assertCount(105, $split['xTrain']);
        $this->assertCount(45, $split['xTest']);

        // Java 版・Scala 版・Clojure 版・Elixir 版と一致する。
        $values = array_column($split['xTrain'], 'がく片長さ');
        $this->assertEqualsWithDelta(
            0.4215384615384616,
            array_sum($values) / count($values),
            1e-15,
        );
    }

    #[Group('data')]
    #[TestDox('実データの前処理の結果を表示する')]
    public function test実データの前処理の結果を表示する(): void
    {
        if (!Dataset::exists('iris.csv')) {
            $this->markTestSkipped('学習データがありません: ' . Dataset::path('iris.csv'));
        }

        $output = Chapter02::run();

        $this->assertStringContainsString('データ件数: 150', $output);
        $this->assertStringContainsString('訓練データ: 105 件, テストデータ: 45 件', $output);
    }
}
